## **Project name**   - Uber Supply_demand Gaps

##### **Project Type**    - EDA
##### **Contribution**    - Individual
##### **Name**            - Kashish Chaudhary

# **Project Summary -**

This project aimed to perform a comprehensive analysis of Uber ride request data to identify supply-demand gaps, delays, cancellations, and behavioral patterns. The process was executed using three powerful tools: Python (Pandas) for data cleaning, Microsoft Excel for interactive dashboards, and SQL for structured query-based insights. Each stage was critical to building a polished, insightful, and user-friendly visualization product.

1. Data Cleaning and Structuring with Pandas The initial raw dataset contained mixed timestamp formats, null values, and overlapping data fields. Using Pandas in Python, the data underwent structured cleaning:
The CSV was read using pandas.read_csv().

Combined datetime columns (request_timestamp, drop_timestamp) were converted into Python datetime objects.

These fields were split into individual components: request_date, drop_date, request_time, and drop_time, to allow more granular time-based filtering and analysis.

A new column trip_duration_min was calculated to quantify ride durations.

Null values were intentionally preserved in relevant rows to allow examination of when supply was not available or trips were canceled.

After cleaning, the dataset was saved in both sorted and unsorted versions by request_id, offering two perspectives for dashboard analysis.

2. Dashboard Design and Visualization in Excel The cleaned file was then imported into Excel, where the core dashboard was created using Pivot Tables, Pivot Charts, Slicers, and KPI Cards:
A new column request_in_Hr was created using Excel formulas to convert request_time to a 12-hour AM/PM format, which allowed cleaner time segmentation in visuals.

-> KPI cards were created to highlight core metrics:

*Total Requests

*Completed Trips

*Cancellation Percentage

*Average Trip Duration

*Total Drivers

-> Pivot Tables were set up to analyze:

*Status breakdown by pickup point

*Trip trends by request and drop date

*Hourly patterns (12-hour format)

-> Charts used included:

Bar chart: request vs drop date

Pie chart: ride status distribution

Line chart: request trend over time

->Slicers were added for dynamic filtering across key fields like request_date, driver_id, pickup_point, and status.

The final dashboard followed a dark-blue and gold theme, balancing professional visual design with data readability.

3. Insights with SQL in MySQL Workbench To validate and extend the analysis, the dataset was imported into MySQL Workbench:
A clean database and uber_data table were created with appropriate data types (DATETIME, DATE, TIME, VARCHAR, DECIMAL, etc.).

SQL queries were written to:

Count total requests, completed trips, and cancellations

Analyze trip patterns by date and pickup point

Determine busiest hours and no-driver-available periods

Support cross-validation of dashboard KPIs

Conclusion This project demonstrates a complete, cross-functional data analysis workflow—from cleaning with Pandas, dashboarding with Excel, to querying with SQL. The final deliverable was an interactive and well-designed Excel dashboard supported by reliable backend data processing and SQL insight generation. This approach offers clear visibility into Uber’s operational gaps and serves as a valuable analytical template for ride-share performance evaluation.

# **GitHub Link -** 
https://github.com/kashishch28/DA-Projects

# **Problem Statement**


Uber often faces supply-demand mismatches, especially during peak hours, resulting in ride cancellations and unfulfilled requests. Identifying when and where these gaps occur is essential to improve service efficiency and user satisfaction. This project analyzes Uber ride data to uncover trends in request volume, driver availability, and ride status. Using Python for data cleaning, Excel for dashboard creation, and SQL for deeper analysis, the goal is to highlight patterns in supply shortfalls and provide actionable insights. The final outcome is a professional dashboard and a set of SQL queries to support better operational planning.




# ***Let's Begin !***

## ***1. Know Your Data***


### Import Libraries


In [15]:
# Import Libraries
import pandas as pd


### Dataset Loading

In [5]:
# Load Dataset
df = pd.read_csv(r"C:\Users\kashi\Downloads\Uber Request Data.csv")


### Dataset First View

In [6]:
df

Request id Pickup point  Driver id             Status Request timestamp  \
0            619      Airport        1.0     Trip Completed  11-07-2016 11:51   
1            867      Airport        1.0     Trip Completed  11-07-2016 17:57   
2           1807         City        1.0     Trip Completed  12-07-2016 09:17   
3           2532      Airport        1.0     Trip Completed  12-07-2016 21:08   
4           3112         City        1.0     Trip Completed  13-07-2016 08:33   
...          ...          ...        ...                ...               ...   
6740        6745         City        NaN  No Cars Available  15-07-2016 23:49   
6741        6752      Airport        NaN  No Cars Available  15-07-2016 23:50   
6742        6751         City        NaN  No Cars Available  15-07-2016 23:52   
6743        6754         City        NaN  No Cars Available  15-07-2016 23:54   
6744        6753      Airport        NaN  No Cars Available  15-07-2016 23:55   

        Drop timestamp  
0     11-07-2016 13:00  
1     11-07-2016 18:47  
2     12-07-2016 09:58  
3     12-07-2016 22:03  
4     13-07-2016 09:25  
...                ...  
6740               NaN  
6741               NaN  
6742               NaN  
6743               NaN  
6744               NaN  

[6745 rows x 6 columns]

### Dataset Rows & Columns count

In [7]:
df.shape

(6745, 6)

### Dataset Information

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6745 entries, 0 to 6744
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Request id         6745 non-null   int64  
 1   Pickup point       6745 non-null   object 
 2   Driver id          4095 non-null   float64
 3   Status             6745 non-null   object 
 4   Request timestamp  6745 non-null   object 
 5   Drop timestamp     2831 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 316.3+ KB


### Missing Values/Null Values

In [12]:
df.isnull().sum()

Request id              0
Pickup point            0
Driver id            2650
Status                  0
Request timestamp       0
Drop timestamp       3914
dtype: int64

### What did you know about your dataset?

#### The following are the major points about the dataset:-

1. The dataset contains 6 columns and 6745 rows.
2. There are no duplicate values in the dataset.

## ***2. Understanding Your Variables***

In [16]:
# Dataset Columns
df.columns

Index(['Request id', 'Pickup point', 'Driver id', 'Status',
       'Request timestamp', 'Drop timestamp'],
      dtype='object')

In [18]:
# Dataset Describe
df.describe()

Request id    Driver id
count  6745.000000  4095.000000
mean   3384.644922   149.501343
std    1955.099667    86.051994
min       1.000000     1.000000
25%    1691.000000    75.000000
50%    3387.000000   149.000000
75%    5080.000000   224.000000
max    6766.000000   300.000000

### Variables Description

#### The following are the variables with their description:-

#### *Column names---------------------------Description*

1. request_id->Customer Unique id who requested for the cab service through app.
2. Pickup point->Pickup location to pickup the customer.
3. Driver id->Unique Id of driver who is driving the cab.
4. Status->status of trip that if it got completed, cancelled or no cabs are available.
5. Request timestamp->Request time and date when the customer requested for the cab service.
6. Drop timestamp->Drop time and date when the customer is dropped to their respective location.
##### The above are the descriptions for the variables present in the dataset.

### Check Unique Values for each variable.

In [24]:
for x in df.columns:
    print(f"{x} - {df[x].nunique()}")

Request id - 6745
Pickup point - 2
Driver id - 300
Status - 3
Request timestamp - 4016
Drop timestamp - 2282


## 3. ***Data Wrangling***

In [26]:
df.isnull().sum()

Request id              0
Pickup point            0
Driver id            2650
Status                  0
Request timestamp       0
Drop timestamp       3914
dtype: int64

In [27]:
df.fillna({"Driver id": "Not Available"},inplace=True)

In [28]:

df.fillna({"Drop timestamp": "Not Available"},inplace=True)

In [29]:

# checking that the missing values are present or removed
df.isnull().sum()

Request id           0
Pickup point         0
Driver id            0
Status               0
Request timestamp    0
Drop timestamp       0
dtype: int64

### What all manipulations have you done and insights you found?

The following manipulation we have done to our dataset:-

As there were no duplicates so no issue of removing it.
There are total 2 columns having missing values which we replaced with Not Available value.
Driver id-----2650

Drop timestamp-----3914

## ***4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables***

For Visualization excel sheets is attached

# **Conclusion**

### **Conclusion**

Through this project, we performed an end-to-end analysis of Uber ride request data using **Pandas (Python)** for data cleaning, **Excel** for dashboard visualization, and **SQL** for backend data validation.

In **Pandas**, we handled time formatting, removed redundant columns, split timestamps into date/time, and calculated trip durations. This enabled structured and accurate analysis downstream.

In **Excel**, we created an interactive dashboard featuring bar, pie, and line charts, slicers for filtering, and KPI cards. These visuals helped uncover key insights such as:

* High cancellation rates during peak hours
* Supply shortages, especially at the Airport pickup point
* Specific times of day when "No Cars Available" was common
* Trends in completed vs. unfulfilled trips

In **SQL**, we ran queries to validate Excel findings, calculate aggregates, and confirm logic behind our KPIs—such as total trips, completion rate, and hourly patterns.

Together, these tools revealed the root causes of supply-demand mismatches and offered actionable insights for operational improvements. The combined use of Pandas, Excel, and SQL provided a robust analytical workflow, enabling both visual storytelling and structured data validation.
